schema/gen/go/genMap.go

```diff
@@ -0,0 +1,631 @@
+package gengo
+
+import (
+	"io"
+
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/ipld/go-ipld-prime/schema/gen/go/mixins"
+)
+
+type mapGenerator struct {
+	AdjCfg *AdjunctCfg
+	mixins.MapTraits
+	PkgName string
+	Type    schema.TypeMap
+}
+
+// --- native content and specializations --->
+
+func (g mapGenerator) EmitNativeType(w io.Writer) {
+	// Maps do double bookkeeping.
+	// - 'm' is used for quick lookup.
+	// - 't' is used for both for order maintainence, and for allocation amortization for both keys and values.
+	// Note that the key in 'm' is *not* a pointer.
+	// The value in 'm' is a pointer into 't' (except when it's a maybe; maybes are already pointers).
+	doTemplate(`
+		type _{{ .Type | TypeSymbol }} struct {
+			m map[_{{ .Type.KeyType | TypeSymbol }}]{{if .Type.ValueIsNullable }}Maybe{{else}}*_{{end}}{{ .Type.ValueType | TypeSymbol }}
+			t []_{{ .Type | TypeSymbol }}__entry
+		}
+		type {{ .Type | TypeSymbol }} = *_{{ .Type | TypeSymbol }}
+	`, w, g.AdjCfg, g)
+	// - address of 'k' is used when we return keys as nodes, such as in iterators.
+	//    Having these in the 't' slice above amortizes moving all of them to heap at once,
+	//     which makes iterators that have to return them as an interface much (much) lower cost -- no 'runtime.conv*' pain.
+	// - address of 'v' is used in map values, to return, and of course also in iterators.
+	doTemplate(`
+		type _{{ .Type | TypeSymbol }}__entry struct {
+			k _{{ .Type.KeyType | TypeSymbol }}
+			v _{{ .Type.ValueType | TypeSymbol }}{{if .Type.ValueIsNullable }}__Maybe{{end}}
+		}
+	`, w, g.AdjCfg, g)
+}
+
+func (g mapGenerator) EmitNativeAccessors(w io.Writer) {
+	// Generate a speciated Lookup as well as LookupMaybe method.
+	// The LookupMaybe method is needed if the map value is nullable and you're going to distinguish nulls
+	//  (and may also be convenient if you would rather handle Maybe_Absent than an error for not-found).
+	// The Lookup method works fine if the map value isn't nullable
+	//  (and should be preferred in that case, because boxing something into a maybe when it wasn't already stored that way costs an alloc(!),
+	//   and may additionally incur a memcpy if the maybe for the value type doesn't use pointers internally).
+	// REVIEW: is there a way we can make this less twisty?  it is VERY unfortunate if the user has to know what sort of map it is to know which method to prefer.
+	//  Maybe the Lookup method on maps that have nullable values should just always have a MaybeT return type?
+	//   But then this means the Lookup method doesn't "need" an error as part of its return signiture, which just shuffles differences around.
+	doTemplate(`
```

Comment:
Dear Reviewer, may I direct your attention to this area especially?  There's quite a lot of ergonomics questions here.  I really haven't sorted out what I even *want* this to look like, and there seems to be a tug-o-war between pleasantness/consistency versus making it obvious what's performant and natural to prefer the usage that is fastest.